easy_kclpy/kinesis_processor.py

```diff
@@ -106,40 +75,78 @@ def checkpoint(self, checkpointer, sequence_number=None, sub_sequence_number=Non
                 elif 'InvalidStateException' == e.value:
                     log.error('MultiLangDaemon reported an invalid state while checkpointing.')
                 else:  # Some other error
-                    log.error('Encountered an error while checkpointing, error was {e}.'.format(e=e))
+                    log.error('Encountered an error while checkpointing: {e}.'.format(e=e))
             sleep(self.checkpoint_error_sleep_seconds)
 
-    def shutdown(self, shutdown_input):
-        """
-        Called by a KCLProcess instance to indicate that this record processor should shutdown. After this is called,
-        there will be no more calls to any other methods of this record processor.
+        self.after_checkpoint(sequence_number, sub_sequence_number)
 
-        As part of the shutdown process you must inspect :attr:`amazon_kclpy.messages.ShutdownInput.reason` to
-        determine the steps to take.
+    def process_record(self, data, partition_key, sequence_number, sub_sequence_number, approximate_arrival_timestamp):
+        """
+        Called for each record that is passed to process_records.
 
-            * Shutdown Reason ZOMBIE:
-                **ATTEMPTING TO CHECKPOINT ONCE A LEASE IS LOST WILL FAIL**
+        :param str data: The blob of data that was contained in the record.
+        :param str partition_key: The key associated with this recod.
+        :param int sequence_number: The sequence number associated with this record.
+        :param int sub_sequence_number: the sub sequence number associated with this record.
+        :param approximate_arrival_timestamp: the time the record arrived
+        """
+        raise NotImplementedError
 
-                A record processor will be shutdown if it loses its lease.  In this case the KCL will terminate the
-                record processor.  It is not possible to checkpoint once a record processor has lost its lease.
-            * Shutdown Reason TERMINATE:
-                **THE RECORD PROCESSOR MUST CHECKPOINT OR THE KCL WILL BE UNABLE TO PROGRESS**
+    def should_update_sequence(self, sequence_number, sub_sequence_number):
```

Comment:
Is this from a reference implementation?

Reply:
>80% of the code is from here:
https://github.com/awslabs/amazon-kinesis-client-python